common/cluster/dynamic_metadata.go

```diff
@@ -0,0 +1,502 @@
+
+	result := make(map[string]ClusterInformation, len(d.clusterInfo))
+	for k, v := range d.clusterInfo {
+		result[k] = v
+	}
+	return result
+}
+
+func (d *dynamicMetadataImpl) ClusterNameForFailoverVersion(failoverVersion int64) string {
+	if failoverVersion == common.EmptyVersion {
+		return d.currentClusterName
+	}
+
+	initialFailoverVersion := failoverVersion % d.failoverVersionIncrement
+	// Failover version starts with 1.  Zero is an invalid value for failover version
+	if initialFailoverVersion == common.EmptyVersion {
+		initialFailoverVersion = d.failoverVersionIncrement
+	}
+
+	d.clusterLock.RLock()
+	defer d.clusterLock.RUnlock()
+	clusterName, ok := d.versionToClusterName[initialFailoverVersion]
+	if !ok {
+		panic(fmt.Sprintf(
+			"Unknown initial failover version %v with given cluster initial failover version map: %v and failover version increment %v.",
+			initialFailoverVersion,
+			d.clusterInfo,
+			d.failoverVersionIncrement,
+		))
+	}
+	return clusterName
+}
+
+func (d *dynamicMetadataImpl) GetFailoverVersionIncrement() int64 {
+	return d.failoverVersionIncrement
+}
+
+func (d *dynamicMetadataImpl) RegisterMetadataChangeCallback(callbackId string, cb CallbackFn) {
+	d.clusterCallbackLock.Lock()
+	d.clusterChangeCallback[callbackId] = cb
+	d.clusterCallbackLock.Unlock()
+
+	oldEntries := make(map[string]*ClusterInformation)
+	newEntries := make(map[string]*ClusterInformation)
+	d.clusterLock.RLock()
+	for clusterName, clusterInfo := range d.clusterInfo {
+		oldEntries[clusterName] = nil
+		newEntries[clusterName] = &ClusterInformation{
+			Enabled:                clusterInfo.Enabled,
+			InitialFailoverVersion: clusterInfo.InitialFailoverVersion,
+			RPCAddress:             clusterInfo.RPCAddress,
+			version:                clusterInfo.version,
+		}
+	}
+	d.clusterLock.RUnlock()
+	cb(oldEntries, newEntries)
+}
+
+func (d *dynamicMetadataImpl) UnRegisterMetadataChangeCallback(callbackId string) {
+	d.clusterCallbackLock.Lock()
+	delete(d.clusterChangeCallback, callbackId)
+	d.clusterCallbackLock.Unlock()
+}
+
+func (d *dynamicMetadataImpl) refreshLoop(ctx context.Context) error {
+	timer := time.NewTicker(refreshInterval)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-timer.C:
+			for err := d.refreshClusterMetadata(ctx); err != nil; err = d.refreshClusterMetadata(ctx) {
+				select {
+				case <-ctx.Done():
+					return nil
+				default:
+					d.logger.Error("Error refreshing remote cluster metadata", tag.Error(err))
+					select {
+					case <-time.After(refreshFailureInterval):
+					case <-ctx.Done():
+						return nil
+					}
+				}
+			}
+		}
+	}
+}
+
+func (d *dynamicMetadataImpl) refreshClusterMetadata(_ context.Context) error {
+	clusterMetadataMap, err := d.listAllClusterMetadata()
+	if err != nil {
+		return err
+	}
+
+	oldEntries := make(map[string]*ClusterInformation)
+	newEntries := make(map[string]*ClusterInformation)
+
+	clusterInfoMap := d.GetAllClusterInfo()
```

Comment:
it is confusing to see listAllClusterMetadata() and GetAllClusterInfo(), is list from db and Get from memory cache?

Reply:
Yes. I will update the private method name